cmd/snap/cmd_connections.go

```diff
@@ -0,0 +1,132 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/snapcore/snapd/i18n"
+
+	"github.com/jessevdk/go-flags"
+)
+
+type cmdConnections struct {
+	clientMixin
+	All         bool `short:"a"`
+	Positionals struct {
+		Query interfacesSlotOrPlugSpec `skip-help:"true"`
+	} `positional-args:"true"`
+}
+
+var shortConnectionsHelp = i18n.G("List interface connections")
+var longConnectionsHelp = i18n.G(`
+List connections between slots and plugs. When passed an optional
+snap name, lists connection only for that snap.
+
+Pass -a to list slots/plugs that are not connected to anything.
+`)
+
+func init() {
+	addCommand("connections", shortConnectionsHelp, longConnectionsHelp, func() flags.Commander {
+		return &cmdConnections{}
+	}, map[string]string{
+		"a": i18n.G("Show all connections"),
+	}, []argDesc{{
+		// TRANSLATORS: This needs to be wrapped in <>s.
+		name: i18n.G("<snap>"),
+		// TRANSLATORS: This should not start with a lowercase letter.
+		desc: i18n.G("Constrain listing to a specific snap"),
+	}})
+}
+
+func endpoint(snap, name string) string {
+	if snap == "core" || snap == "snapd" || snap == "system" {
+		return ":" + name
+	}
+	return snap + ":" + name
+}
+
+func wantedSnapMatches(name, wanted string) bool {
+	if wanted == "system" {
+		switch name {
+		case "core", "snapd", "system":
+			return true
+		default:
+			return false
+		}
+	}
+	return wanted == name
+}
+
+func (x *cmdConnections) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+	if x.Positionals.Query.Name != "" {
+		return errors.New("filtering by slot/plug name is not supported")
+	}
+
+	ifaces, err := x.client.Connections()
```

Comment:
ugh, wonder why we called it Connections in client for the current API, we might consider renaming that if we introduce a new one